src/cc/BPFTable.h

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright (c) 2016 Facebook, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <exception>
+#include <memory>
+#include <string>
+#include <utility>
+#include <vector>
+
+#include "bpf_module.h"
+
+namespace ebpf {
+
+class BPFTableBase {
+ public:
+  size_t max_size() {
+    return max_size_;
+  }
+
+ protected:
+  BPFTableBase(std::shared_ptr<BPFModule> bpf_module, const std::string& name);
+
+  bool lookup(void* key, void* value);
+  bool next(void* key, void* next_key);
+
+  const std::string name_;
+  std::shared_ptr<BPFModule> bpf_module_;
+  size_t id_;
+  int fd_;
+  size_t max_size_;
+};
+
+template<class KeyType, class ValueType>
+class BPFHashTable : protected BPFTableBase {
+ public:
+  BPFHashTable(std::shared_ptr<BPFModule> bpf_module, const std::string& name)
+    : BPFTableBase(bpf_module, name) {}
+
+  ValueType get_value(const KeyType& key) {
+    ValueType res;
+    if (!lookup(
+      const_cast<void*>(static_cast<const void*>(&key)),
+      static_cast<void*>(&res)
+    ))
+      throw std::invalid_argument("Key does not exist in the table");
+    return res;
+  }
+
+  ValueType operator[](const KeyType& key) {
+    return get_value(key);
+  }
+
+  std::vector<std::pair<KeyType, ValueType>> get_table_offline() {
```

Comment:
such 'return all entries' api is dangerous. I don't think there is ever a valid use case to do so, yet it's trivial to abuse it.